### PR TITLE
WebSocket improvements, SSL server, custom headers.

### LINKED ISSUE #27129: WebSocketClient has no support for custom HTTP headers
**Godot version:** 3.1

**OS/device including version:** Not relevant

**Issue description:**

The ```WebSocketClient``` does not provide an option for custom HTTP headers on ```connect_to_url```, like for example ```HTTPRequest```. This is at least relevant for the usual HTTP authorization flows to work (Basic, JWT, etc.), before even upgrading the request.

**Steps to reproduce:** See API definition.

**Minimal reproduction project:** Not relevant.

### PR DESCRIPTION
**Pull request by Faless**
In this PR:

- Reimplement `WebSocketPeer` `get_connected_port` and `get_connected_host` (as always, not supported in HTML5)
- Allow specifying the X509 certificate to be trusted by `WebSocketClient` when using SSL (an alternative to use project-wise certs).
- Implement `WebSocketServer` SSL support via new properties `private_key`, `ssl_certificate`.
- Allow using passing custom headers to `WebSocketClient.connect_to_url` (Closes #27129 , Closes #25999 )

Document all the new goodies.

### CODE REVIEW

File: modules/websocket/doc_classes/WebSocketServer.xml [hunk 1]
```diff
@@ -82,6 +82,17 @@
 			</description>
 		</method>
 	</methods>
+	<members>
+		<member name="ca_chain" type="X509Certificate" setter="set_ca_chain" getter="get_ca_chain">
+			When using SSL (see [member private_key] and [member ssl_certificate]), you can set this to a valid [X509Certificate] to be provided as additional CA chain information during the SSL handshake.
+		</member>
+		<member name="private_key" type="CryptoKey" setter="set_private_key" getter="get_private_key">
+			When set to a valid [CryptoKey] (along with [member ssl_certificate]) will cause the server to require SSL instead of regular TCP (i.e. the `wss://` protocol).
```
Comment (akien-mga) [inline modules/websocket/doc_classes/WebSocketServer.xml]: Should be `[code]wss://[/code]`.

File: modules/websocket/doc_classes/WebSocketServer.xml [hunk 2]
```diff
@@ -82,6 +82,17 @@
...
+		</member>
+		<member name="ssl_certificate" type="X509Certificate" setter="set_ssl_certificate" getter="get_ssl_certificate">
+			When set to a valid [X509Certificate] (along with [member private_key]) will cause the server to require SSL instead of regular TCP (i.e. the `wss://` protocol).
```
Comment (akien-mga) [inline modules/websocket/doc_classes/WebSocketServer.xml]: Same here.

File: modules/websocket/doc_classes/WebSocketClient.xml [hunk 3]
```diff
@@ -38,8 +41,25 @@
 				Disconnects this client from the connected host. See [method WebSocketPeer.close] for more information.
 			</description>
 		</method>
+		<method name="get_connected_host" qualifiers="const">
+			<return type="String">
+			</return>
+			<description>
+				Return the IP address of the currently connected host.
```
Comment (akien-mga) [inline modules/websocket/doc_classes/WebSocketClient.xml]: Should these two also mention that they're not supported on HTML5?

File: modules/websocket/doc_classes/WebSocketServer.xml (hunk 1 above)
Comment (akien-mga) [reply modules/websocket/doc_classes/WebSocketServer.xml]: I'll do the fixup in a post-merge commit, I have a few other doc things to check in.

**Comment by akien-mga** [comment]
Thanks!

File: modules/websocket/doc_classes/WebSocketServer.xml (hunk 1 above)
Comment (Faless) [reply modules/websocket/doc_classes/WebSocketServer.xml]: Oh, thanks, sorry about the delay

**Pull request merged**